bevy_lint/src/lints/borrow_of_reborrowable.rs

```diff
@@ -0,0 +1,224 @@
+//! Checks for function parameters that take a mutable reference to a
+//! re-borrwable type.
+//!
+//! # Motivation
+//!
+//! Mutable references to re-borrwable types can almost always be readily
+//! converted back to an owned instance of itself, albeit with an appropriately
+//! shorter lifetime.
+//!
+//! The only time this isn't true is when the function returns referenced data
+//! that is bound to the mutable reference of the re-borrowable type.
+//!
+//! # Example
+//!
+//! ```
+//! # use bevy::prelude::*;
+//! fn system(mut commands: Commands) {
+//!   helper_function(&mut commands);
+//! }
+//!
+//! fn helper_function(commands: &mut Commands) {
+//!   // ...
+//! }
+//! ```
+//!
+//! Use instead:
+//!
+//! ```
+//! # use bevy::prelude::*;
+//! fn system(mut commands: Commands) {
+//!   helper_function(commands.reborrow());
+//! }
+//!
+//! fn helper_function(mut commands: Commands) {
+//!   // ...
+//! }
+//! ```
+
+use std::ops::ControlFlow;
+
+use crate::declare_bevy_lint;
+use clippy_utils::{diagnostics::span_lint_and_sugg, ty::match_type};
+use rustc_errors::Applicability;
+use rustc_hir::{intravisit::FnKind, Body, FnDecl, Mutability};
+use rustc_lint::{LateContext, LateLintPass, Lint};
+use rustc_middle::{
+    ty::Interner,
+    ty::{Ty, TyKind, TypeVisitable, TypeVisitor},
+};
+use rustc_session::declare_lint_pass;
+use rustc_span::{def_id::LocalDefId, symbol::Ident, Span};
+
+declare_bevy_lint! {
+    pub BORROW_OF_COMMANDS,
+    PEDANTIC,
+    "parameter takes a mutable reference to `Commands` or `EntityCommands` instead of a re-borrowed instance",
+}
+
+declare_bevy_lint! {
+    pub BORROW_OF_RESOURCE,
+    PEDANTIC,
+    "parameter takes a mutable reference to `ResMut` or `NonSendMut` instead of a re-borrowed instance",
+}
+
+declare_bevy_lint! {
+    pub BORROW_OF_QUERY,
+    PEDANTIC,
+    "parameter takes a mutable reference to `Query` instead of a re-borrowed instance",
+}
```

Comment:
Do you have a specific reason for splitting these into 3 different lints instead of one? Is there going to be a time where a user will want to toggle on only a subset of these?

Reply:
No reason other than I thought it might be nice to provide the flexibility. I can combine them into a single lint if that's preferred!

Reply:
I think I'd combine them in this case. I realize that [`panicking_methods`](https://github.com/TheBevyFlock/bevy_cli/blob/2480d608b4cb527a01c463b89b620e626f7108c7/bevy_lint/src/lints/panicking_methods.rs#L91-L101) has two different lints, but I've been thinking about combining those as well. (In the future when we have the ability to configure lints, I definitely want them combined!)